include/ginkgo/core/base/segmented_array.hpp

```diff
@@ -0,0 +1,225 @@
+
+    /**
+     * Creates an uninitialized segmented array from offsets.
+     *
+     * @param offsets  the index offsets for each segment, and the total size of
+     *                 the buffer as last element
+     */
+    static segmented_array create_from_offsets(gko::array<int64> offsets);
+
+    /**
+     * Creates a segmented array from a flat buffer and offsets.
+     *
+     * @param buffer  the flat buffer whose size has to match the last element
+     *                of offsets
+     * @param offsets  the index offsets for each segment, and the total size of
+     *                 the buffer as last element
+     */
+    static segmented_array create_from_offsets(gko::array<T> buffer,
+                                               gko::array<int64> offsets);
+
+    /**
+     * Copies a segmented array to a different executor.
+     *
+     * @param exec  the executor to copy to
+     * @param other  the segmented array to copy from
+     */
+    segmented_array(std::shared_ptr<const Executor> exec,
+                    const segmented_array& other);
+
+    /**
+     * Moves a segmented array to a different executor.
+     *
+     * @param exec  the executor to move to
+     * @param other  the segmented array to move from
+     */
+    segmented_array(std::shared_ptr<const Executor> exec,
+                    segmented_array&& other);
+
+    segmented_array(const segmented_array& other);
+
+    segmented_array(segmented_array&& other) noexcept(false);
+
+    segmented_array& operator=(const segmented_array& other);
+
+    segmented_array& operator=(segmented_array&&) noexcept(false);
+
+    /**
+     * Get the total size of the stored buffer.
+     *
+     * @return  the total size of the stored buffer.
+     */
+    size_type get_size() const;
+
+    /**
+     * Get the number of segments.
+     *
+     * @return  the number of segments
+     */
+    size_type get_segment_count() const;
+
+    /**
+     * Access to the flat buffer.
+     *
+     * @return  the flat buffer
+     */
+    T* get_flat_data();
+
+    /**
+     * Const-access to the flat buffer
+     *
+     * @return  the flat buffer
+     */
+    const T* get_const_flat_data() const;
+
+    /**
+     * Access to the segment offsets.
+     *
+     * @return  the segment offsets
+     */
+    const gko::array<int64>& get_offsets() const;
+
+    /**
+     * Access the executor.
+     *
+     * @return  the executor
+     */
+    std::shared_ptr<const Executor> get_executor() const;
+
+private:
+    gko::array<T> buffer_;
+    gko::array<int64> offsets_;
+};
+
+template <typename T>
+size_type segmented_array<T>::get_size() const
+{
+    return buffer_.get_size();
+}
+
+
+template <typename T>
+size_type segmented_array<T>::get_segment_count() const
+{
+    return offsets_.get_size() ? offsets_.get_size() - 1 : 0;
+}
+
+
+template <typename T>
+T* segmented_array<T>::get_flat_data()
+{
+    return buffer_.get_data();
```

Comment:
You can move this directly within the class, right ?

Reply:
I can actually move them into the .cpp. I did that, but I won't wait for the full CI pipeline to merge it.